### DISCUSSION TITLE
Method pattern matching?

### DISCUSSION THREAD

**Discussion by dcuccia** [dotnet/roslyn#60206] [Ideas]
Has pattern matching for methods or lambda function properties been discussed previously? The structural typing now available via the [various patterns](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns) is fantastic, but the functional programmer in me really would like the language to help me program against _capabilities_. Reading [today's post](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-7-preview-2/) about the latest example of convention-based feature support (i.e. HTTPS array binding works for any class with a proper `TryParse` method) makes me wish this were a general capability for us mere mortals, vs specialized reflection code (or Roslyn generator code) that is more cumbersome to write and support.

**Comment by Richiban** [comment]
Can you write a couple of small examples that will help illustrate what you'd like to be able to do? Right now I'm finding it a bit hard to imagine the new feature that you're asking for.

**Reply by dcuccia** [reply]
@Richiban great suggestion, thanks. Off the top of my head...

Enumeration based on shape...
```csharp
IEnumerable<T>? MaybeGetEnumerable<T>(object o)
{
    if(o is { bool MoveNext() } and { T get_Current() } myStructuralEnumerable)
    {
        while(myStructuralEnumerable.MoveNext())
        {
            yield return myStructuralEnumerable.get_Current();
        }
    }
    return null;
}
```

Optional Dispose...
```csharp
interface ICustomTask{ void DoWork(); }
void DoWorkAndDisposeIfNecessary(ICustomTask task)
{
    try
    {
        task.DoWork();
    }
    finally
    {
        if(task is { Dispose() } myDisposableTask)
        {
            myDisposableTask.Dispose();
        }
    }
}
```

Example to elaborate on my point about the .Net 7 [TryParse example](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-7-preview-2/#binding-arrays-and-stringvalues-from-headers-and-query-strings-in-minimal-apis)
```csharp
class ConvolutedMinimalApiQueryStringParserExample
{
    class MyCustomParser
    {
        Type ParseType { get; }
        bool TryParse(string s, out T myValue) { /* ... */ }
    }

    private IEnumerable<object> _registeredParsersForQueryStringResolution = new [] { new MyCustomParser() };
    
    public object ResolveQueryString(params string[] elements)
    {
        foreach(var parser in _registeredParsersForQueryStringResolution)
        {
            if(parser is { Type get_ParseType() } parserWithParseType)
            {
                var type = parserWithParseType.get_Type();
                // shorthand...use reflection/code-gen to call generic method
                TValue[]? result = ResolveQueryStringForComplexType<type>(parser, elements);
                if (result is not null)
                    return result;
            }
        }
        return null;
    }

    private TValue[]? ResolveQueryStringForComplexType<TValue>(object parser, params string[] elements)
    {
        if (parser is T? TryParse(string, out T) myComplexTypeParser && myComplexTypeParser.TryParse(elements?[0], out T myValue))
        {
            var myValues = (T[])Array.CreateInstance(myValue.GetType(), elements.Length);
            myValues[0] = myValue;
            for(int i = 1; i < myValues.Length; i++)
            {
                T myLocalValue;
                if(myComplexTypeParser.TryParse(elements[i], out T myLocalValue))
                    myValues[i] = myLocalValue;
            }
            return myValues;
        }
        return null;
    }
}
```
The above is totally convoluted, particularly since the new query string resolution works on a static TryParse, and the is/as matching works on instances. Of course, if there was an ability to identify/select for static methods in an instance's class, that would be even cooler.

I realize that today's property patterns operate/switch on runtime values, not necessarily on the shape (i.e. property name) alone, so I'm probably suggesting something crazy here. I probably should look up the latest proposals on duck typing etc...I was just heads-down reading the release notes yesterday, and thinking how powerful it would be if we mortal developers had first-class language features to operate on shapes, without always having to take hard dependencies on interfaces. I also appreciate the loosely typed direction this takes the language may not be desirable/idiomatic.

**Reply by dcuccia** [reply]
Another way could be to allow for ad-hoc interfaces to be defined by the consuming code, and use a keyword to denote the runtime association. `if(myObject is extern(IConsumingInterface))` or `if(myObject is unchecked(IConsumingInterface))` haha. That could also bring in the new "static abstract" interface constraints.

**Reply by dcuccia** [reply]
or push that onto the interface, so you could opt-in your existing interfaces to it `extern interface IConsumingInterface{ ... }`, and the code gets cleaner (not sure if that is a mutually-exclusive use of extern as currently used). I'm thinking about how TypeScript works for structural typing.

**Reply by Richiban** [reply]
The sentence in your OP "The structural typing now available via the [various patterns](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns) is fantastic" isn't quite right; patterns don't have anything to do with structural typing; they're all to do with a simultaneous type test and deconstruction into parts.

The closest thing to structural typing currently on the docket would be the shapes proposal https://github.com/dotnet/csharplang/discussions/164 which, although it enjoys a lot of interest from the language design team, is still a very long way off.
